feat: add fallback mechanism for relayer-rpc-node

koonopek opened the pull request.

koonopek commented on packages/on-chain-relayer/src/core/contract-interactions/get-contract.ts:
```diff
@@ -3,14 +3,23 @@ import { abi as priceFeedsAdapterABI } from "../../../artifacts/contracts/price-
+    getSigner(getProvider(index))
+  );
+
+  return contractDecorator.decorate(contract, signers);
```
We are Wrapping (decorating) contract with "fallback logic"

koonopek commented on packages/on-chain-relayer/src/core/contract-interactions/get-contract.ts:
```diff
@@ -3,14 +3,23 @@ import { abi as priceFeedsAdapterABI } from "../../../artifacts/contracts/price-
   const abi = getAbiForAdapter();
-  return new Contract(adapterContractAddress, abi, signer);
+  const contract = new Contract(adapterContractAddress, abi);
+  const signers = config.rpcUrls.map((_rpcUrl: string, index: number) =>
```
Providers will be used in given sequencer if [0] fails then [1] etc.

koonopek commented on packages/on-chain-relayer/src/fallback-contract.ts:
```diff
@@ -0,0 +1,131 @@
+import { ErrorCode } from "@ethersproject/logger";
+import { config } from "./config";
+
+export class FallbackContractDecorator {
```
Same idea as with RedstoneWrapper for contracts (most of the code is coppied).
I could have used Proxy, however I thought that there is no need for reinventing the wheel

koonopek commented:
Will deployed these nodes for testing purposes

cehali commented on packages/on-chain-relayer/deployed-config/arbitrum-goerli-3-nodes.json:
```diff
@@ -0,0 +1,16 @@
+{
```
I think we can update `arbitrum-goerli.json`. We will not keep 4th relayer alive for a long time.

cehali commented on packages/on-chain-relayer/src/fallback-contract.ts:
```diff
@@ -0,0 +1,131 @@
+import { ErrorCode } from "@ethersproject/logger";
+import { config } from "./config";
+
+export class FallbackContractDecorator {
```
I generally like the idea. Also, we want something similar in oracle-nodes and I think it could be an excellent opportunity to extract this to the small package that can be reused in any project and published on npm.

cehali replied:
I want to use it in oracle-node already 😄

koonopek commented on packages/rpc-providers/README.md:
```diff
@@ -0,0 +1,11 @@
+
+**Tested only with JsonRpcProviders**
+
+## ProviderWithFallback
```
For sure need battle testing

koonopek commented on packages/rpc-providers/test/ProviderWithFallback.test.ts:
```diff
@@ -0,0 +1,132 @@
+const TEST_PRIV_KEY =
+  "0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80";
+
+describe("ProviderWithFallback", () => {
```
Pretty hard to test because ethers.provider is proxy, thus it can't be mocked ..

koonopek commented on .github/workflows/on-chain-relayer-tests.yml:
```diff
@@ -14,6 +14,8 @@ jobs:
     steps:
       - uses: actions/checkout@v3
+      - name: Build dependent packages
+        run: cd ../rpc-providers && yarn && yarn build
```
Is there a better way?

koonopek commented on packages/on-chain-relayer/src/core/contract-interactions/get-provider-or-signer.ts:
```diff
@@ -1,16 +1,28 @@
+
+  if (providers.length === 0) {
+    throw new Error("Please provide at least one RPC_URL in RPC_URLS");
+  } else if (providers.length === 1) {
```
Running fallback provider with one provider doesnt have sense

koonopek commented on packages/rpc-providers/contracts/Counter.sol:
```diff
@@ -0,0 +1,25 @@
+
+pragma solidity ^0.8.4;
+
+contract Counter {
```
for testing purposes

koonopek commented:
@cehali @hatskier Okey so I have finally finished this big guy. I think I implemented it 3 times and after every implementation, I have changed my mind, I hope this version is okey.

cehali commented on .github/workflows/on-chain-relayer-tests.yml:
```diff
@@ -14,6 +14,8 @@ jobs:
         working-directory: packages/on-chain-relayer
     steps:
       - uses: actions/checkout@v3
+      - name: Build dependent packages
+        run: cd ../rpc-providers && yarn && yarn build && cd ../on-chain-relayer
```
I think that can be handled by reference in tsconfig file, take a look at `tsconfig.json` in `on-chain-relayer`

koonopek replied:
Tried but still doesn't work without pre-building deps

cehali replied:
But in the current implementation, the relayer depends on `redstone-sdk` and we don't need to build it in github action.

koonopek replied:
Ohhh it was failing cause in package json `tsc -build` instead of `tsc --build`.
But why do we have @redstone-finance/evm-connector as dependency then and not reference?

koonopek replied:
Also had to add tsconfig.build.json to exclude tests :)
Thanks @cehali for validating this!

cehali commented on packages/on-chain-relayer/package.json:
```diff
@@ -15,7 +15,8 @@
     "ethers": "^5.7.2",
-    "toad-scheduler": "^2.2.0"
+    "toad-scheduler": "^2.2.0",
+    "redstone-rpc-providers": "*"
```
This dependency also can be handled by reference in tsconfig file, we use `redstone-sdk` and `redstone-protocol` but we don't have it specified as a dependency.

koonopek replied:
Hmm, so what are the pros of using "reference"? I would stay with explicit adding is a dep, I think that is "yarn workspaces way of doing this"

cehali commented on packages/on-chain-relayer/package.json:
```diff
@@ -34,6 +35,7 @@
     "hardhat": "^2.13.0",
     "hardhat-gas-reporter": "^1.0.8",
     "msw": "^1.1.0",
+    "sinon": "^15.0.3",
```
Change from other PR? 😄

koonopek replied:
No actually using this in tests to mock Provider

cehali commented on packages/on-chain-relayer/src/core/contract-interactions/get-last-round-params.ts:
```diff
@@ -1,11 +1,11 @@
 export const getLastRoundParamsFromContract = async (
   managerContract: Contract
-) => {
+): Promise<{ lastRound: number; lastUpdateTimestamp: number }> => {
```
Please extract the type to separate the variable with the proper name

koonopek replied:
Done

cehali commented on packages/rpc-providers/src/ProviderWithFallback.ts:
```diff
@@ -0,0 +1,255 @@
+
+export const FALLBACK_DEFAULT_CONFIG: ProviderWithFallbackConfig = {
+  unrecoverableErrors: [
+    ErrorCode.NONCE_EXPIRED,
```
I think we could retry nonce error, in most cases it is probably race condition problem

koonopek replied:
That is the case for relayer node, where we actually expect nonce errors.
But that his valid point I will extract it from Default Config, and include in Relayer Code

cehali commented on packages/rpc-providers/test/ProviderWithFallback.test.ts:
```diff
@@ -0,0 +1,180 @@
+    );
+    const fallbackProvider = new ProviderWithFallback([
+      alwaysFailingProvider,
+      // hardhat.ethers.provider,
```
Unnecessary comment?

koonopek replied:
yep

koonopek commented on packages/rpc-providers/src/ProviderWithFallback.ts:
```diff
@@ -0,0 +1,255 @@
+    return nextProviderIndex;
+  }
+
+  sendTransaction(...args: any[]): Promise<TransactionResponse> {
```
I tried, but if we try to do it dynamically we don't get TypeScript support. For example
```ts
interface Hi { hi(): void };
class A implements Hi {
 constructor() {
   this['hi'] = console.log('hi')
 }
}
```
It will fail to compile.

Thus, I have extracted the abstract class which repeats all boilerplate. Maybe it is more readable now WDYT?

Lukasz2891 commented on packages/rpc-providers/src/ProviderWithFallback.ts:
```diff
@@ -0,0 +1,203 @@
+    }
+
+    // if we haven't tried provider for this request
+    const nextProviderIndex =
```
Won't `%` (modulus) work here?

```js
nextProviderIndex = (this.providerIndex + 1) % this.providers.length
```

koonopek replied:
Yep, cleaner!

koonopek commented on packages/rpc-providers/src/ProviderWithFallback.ts:
```diff
@@ -0,0 +1,203 @@
+    }
+
+    this.config.logger.warn(
+      `Provider ${providerName} failed with error: ${error?.code} ${error.message}`
```
Good catch!

koonopek commented on packages/rpc-providers/src/ProviderWithFallback.ts:
```diff
@@ -0,0 +1,203 @@
+  override removeListener(eventName: EventType, listener: Listener): Provider {
+    this.currentProvider.removeListener(eventName, listener);
+    this.updateGlobalListenerAfterRemoval();
+    return this.currentProvider;
```
I have added new lines by hand for functions with multiple lines + return. Changes to eslint should be done in different PR I think because It will be very hard to review this one (we have global eslint setup)

koonopek commented:
Benchmarks on read arbitrum 3 providers:
FALLBACK READ 0: 707.616ms
FALLBACK READ 1: 561.783ms
FALLBACK READ 2: 1.051s
FALLBACK READ 3: 568.889ms
FALLBACK READ 4: 552.365ms
FALLBACK READ 5: 584.257ms
FALLBACK READ 6: 594.999ms
FALLBACK READ 7: 650.666ms
FALLBACK READ 8: 597.99ms
FALLBACK READ 9: 591.665ms
SOLO  READ 0: 296.884ms
SOLO  READ 1: 152.708ms
SOLO  READ 2: 146.817ms
SOLO  READ 3: 147.758ms
SOLO  READ 4: 163.572ms
SOLO  READ 5: 158.537ms
SOLO  READ 6: 149.834ms
SOLO  READ 7: 148.664ms
SOLO  READ 8: 164.003ms
SOLO  READ 9: 182.735ms

koonopek commented on packages/rpc-providers/src/ProviderWithAgreement.ts:
```diff
@@ -0,0 +1,192 @@
+}
+
+/** Only difference from Promise.race is that it replace resolved promises if undefined */
+function smartPromiseRace<T>(promises: Promise<T>[]): {
```
.

koonopek replied:
This one makes us fast, because we don't have to wait for all promises to resolve

hatskier commented on packages/rpc-providers/src/ProviderWithAgreement.ts:
```diff
@@ -0,0 +1,192 @@
+      );
+
+      const result = await theFastestPromise
+        .then((data) => ({
```
In my opinion, it's not really readable to mix async/await syntax with promises chaining (using `then` and `catch` functions). But it's a matter of taste

koonopek replied:
will do

koonopek replied:
done

hatskier commented on packages/rpc-providers/src/ProviderWithAgreement.ts:
```diff
@@ -0,0 +1,192 @@
+    let resolvedCalls = 0;
+
+    while (true) {
+      const { theFastestPromise, notResolvedPromises } = await smartPromiseRace(
```
I think this algorithm can be improved, as currently it can send too many unnecessary requests to the RPC URLs. We can call all the promises one time and just handle responses on demand, and stop handling them when received enough valid and equal responses. We can save responses in a multiset, or a map like this: hash(response.tx.data) => response.tx.data

koonopek replied:
done

koonopek commented on packages/rpc-providers/src/ProviderWithFallback.ts:
```diff
@@ -0,0 +1,219 @@
+      return await (this.currentProvider as any)[fnName](...[...args]);
+    } catch (error: any) {
+      const newProviderIndex = this.electNewProviderOrFail(error, retryNumber);
+      this.useProvider(newProviderIndex);
```
Soultion: isolate provider counting per call. after failure check if index has already changes if so the do nothing otherwise select new provider

koonopek replied:
done

koonopek commented on packages/rpc-providers/README.md:
```diff
@@ -0,0 +1,19 @@
+## ProviderWithAgreement
+
+1. Works like `ProviderWithFallback, with an exception for two methods.
+2. `getBlockNumber` asks all providers for blockNumber and then picks `blockNumber.sort(ASC)[1]`
```
YES, actually this idea come from you I think, that we should take the second lowest block number as current state

koonopek commented on packages/rpc-providers/README.md:
```diff
@@ -0,0 +1,19 @@
+3. `call` ask all providers for result
+   - if at least `N` answers are the same return call result
+   - ignore all errors
+   - as soon as he will receive two matching responses return. (doesn't wait for the rest of providers to finish)
```
Changed in whole README.md

koonopek commented on packages/rpc-providers/test/ProviderWithAgreement.test.ts:
```diff
@@ -0,0 +1,161 @@
+    });
+
+    it("should read from contract", async () => {
+      const counter = contract.connect(signer.connect(providerWithAgreement));
```
done also in ProviderWithFallback

koonopek commented on packages/rpc-providers/src/ProviderWithFallback.ts:
```diff
@@ -0,0 +1,219 @@
+
+export type ProviderWithFallbackConfig = {
+  unrecoverableErrors: ErrorCode[];
+  logger: { info: (message: string) => void; warn: (message: string) => void };
```
Mine, changed it to use https://docs.ethers.org/v5/api/utils/logger/ (don't allow to pass own logger)

koonopek commented on packages/rpc-providers/src/ProviderWithAgreement.ts:
```diff
@@ -0,0 +1,192 @@
+    const agreedResults: CallResult[] = [];
+    let resolvedCalls = 0;
+
+    while (true) {
```
done

koonopek commented on packages/rpc-providers/src/ProviderWithAgreement.ts:
```diff
@@ -0,0 +1,192 @@
+    const blockNumbers: number[] = [];
+
+    // collect block numbers
+    await Promise.all(
```
done

koonopek commented on packages/on-chain-relayer/src/core/contract-interactions/get-last-round-params.ts:
```diff
@@ -1,11 +1,16 @@
+
+export type LastRoundParams = {
+  lastRound: number;
+  lastUpdateTimestamp: number;
```
Resolved all conflicts

hatskier commented on packages/rpc-providers/src/ProviderWithAgreement.ts:
```diff
@@ -0,0 +1,125 @@
+    const errors = [];
+    for (const blockNumberResult of blockNumbersResults) {
+      if (blockNumberResult.status === "fulfilled") {
+        blockNumbers.push(blockNumberResult.value);
```
I think this algorithm can be simplified if we get rid of the array

hatskier replied:
I would also like to discuss the best way to select the right block number, as taking the second smaller one may be not the best option as well

koonopek replied:
That was the option we have chosen when we were discussing implementation, I was proposing median or the lowest block, but you suggested that heuristic to take the second oldest block should be the best. I agreed, and still agree. We didn't wanted to build full quorum provider, we just wanted to have confirmation that at least two nodes agree on the same answer.

koonopek replied:
done

koonopek replied:
In the end I made this configurable and as default function I have proposed to take `floor(provider.length /3 )`, so depending on the providers count: 
```
5 -> blocks[1] (4/5 synced providers)
6 -> blocks[3] (4/6 synced providers)
7 -> blocks[3] (5/7 synced providers)
10 -> blocks[5] (6/10 synced providers)
```
If we would pick median we would eliminate half of the providers on the start, I think this is too restrictive

hatskier commented on packages/rpc-providers/src/ProviderWithAgreement.ts:
```diff
@@ -0,0 +1,125 @@
+  private resolveCallPromises(callPromises: Promise<string>[]) {
+    return new Promise<string>((resolve, reject) => {
+      const errors: Error[] = [];
+      const agreedResults: string[] = [];
```
I think this algorithm can fail in some cases. E.g. we call 10 promises and require at least 3 same responses. The first answer that we receive is `0x1`, but all the rest is `0x2`

hatskier replied:
I think the proper implementation should have a mapping `(resultHash => {result, counter})`

koonopek replied:
Yes we discussed that this is okey. To be naive this way, because we are waiting only for two responses. If not we should always wait for all responses or increase quorum to at least 3 same answers.

cehali commented on packages/on-chain-relayer/src/core/contract-interactions/get-contract.ts:
```diff
@@ -1,17 +1,20 @@
-  const { privateKey, adapterContractAddress } = config;
-  const provider = getProvider();
-  const signer = new Wallet(privateKey, provider);
+export const getAdapterContract = (signerOrProvider?: Signer | Provider) => {
```
Is change is necessary?

koonopek replied:
Such a method is easier to test because we can inject mock provider for example.
Does it break something?

cehali replied:
Noo, just `signerOrProvider ?? getSigner()` seems a little bit strange 😄

cehali commented on packages/on-chain-relayer/src/core/contract-interactions/get-provider-or-signer.ts:
```diff
@@ -1,16 +1,37 @@
 export const getSigner = () => {
-  const provider = getProvider();
-  const signer = new Wallet(config.privateKey, provider);
+  const signer = new Wallet(config.privateKey, getProvider());
```
`export const getSigner = () => new Wallet ...`

cehali commented on packages/rpc-providers/package.json:
```diff
@@ -0,0 +1,31 @@
+    "sinon": "^15.0.3"
+  },
+  "peerDependencies": {
+    "ethers": "5.x.x"
```
You can point the last 5 version

koonopek replied:
It is peerDependecies so it is saying like: I should work with every ethers 5 version. I think there is no added value in narrowing this.

cehali commented on packages/rpc-providers/src/ProviderWithAgreement.ts:
```diff
@@ -0,0 +1,147 @@
+    return blockNumbers[0];
+  }
+
+  return blockNumbers
+    .sort((a, b) => a - b)
+    .at(Math.floor(numberOfProviders / 3)) as number;
```
Maybe we could take the minimal block number?

cehali commented on packages/rpc-providers/src/ProviderWithAgreement.ts:
```diff
@@ -0,0 +1,147 @@
+      let handledResults = 0;
+
+      for (const callPromise of callPromises) {
+        callPromise
+          .then((currentResult) => {
```
Why not try catch finally?

cehali replied:
Ok, I wasn't aware of this. Thanks for explanation 👍

cehali commented on packages/rpc-providers/src/ProviderWithAgreement.ts:
```diff
@@ -0,0 +1,147 @@
+      for (const callPromise of callPromises) {
+        callPromise
+          .then((currentResult) => {
+            const currentResultCount = results.get(currentResult);
```
What could be the value of `currentResult` string? Is it possible that we will have multiple of the same values?

koonopek replied:
Yes this is expected that different providers on the same block height will return same answer.
This will be result of read method on some contract

cehali replied:
Cool, I get it now 👍

cehali commented on packages/rpc-providers/tsconfig.build.json:
```diff
@@ -0,0 +1,22 @@
+{
```
Why we need this file?

koonopek replied:
We don't wan't to include `tests` in `dist` folder, but still want to type check them :)
Also if we try to use typescript `references` without excluding hardhat tests it will fail. too

koonopek commented on packages/rpc-providers/src/ProviderWithFallback.ts:
```diff
@@ -0,0 +1,237 @@
+      return;
+    }
+
+    const nextProviderIndex = (this.providerIndex + 1) % this.providers.length;
```
If the last provider failed we start from the begging of the array, but for the same request we can try every provider only once.

hatskier commented on packages/on-chain-relayer/test/contracts/price-feeds/old/PriceFeed.test.tes:
```diff
@@ -82,7 +82,9 @@ describe("PriceFeed", () => {
       expect(roundData.answer).to.be.equal(167099000000);
       expect(roundData.roundId).to.be.equal(roundId);
-      expect(roundData.updatedAt).to.be.equal(expectedTimestamp);
+      expect(
```
I think we can remove this file xD

koonopek replied:
We have a lot of files like this, I think they should be removed in separate PR

hatskier commented on packages/rpc-providers/README.md:
```diff
@@ -0,0 +1,19 @@
+3. `call` ask all providers for result
+   - if at least `N` answers are the same, it return call result
+   - ignore all errors
+   - as soon as it will receive two matching responses returns. (doesn't wait for the rest of the providers to finish)
```
Not two matching responses, but N matching responses

koonopek replied:
good

hatskier commented on packages/rpc-providers/src/ProviderWithAgreement.ts:
```diff
@@ -0,0 +1,147 @@
+      this.agreementConfig.numberOfNodesWhichHaveToAgree > this.providers.length
+    ) {
+      throw new Error(
+        "numberOfNodesWhichHaveToAgree should be >= 2 and > then supplied providers count"
```
numberOfNodesWhichHaveToAgree -> numberOfProvidersWhichHaveToAgree?

koonopek replied:
why not

hatskier commented on packages/rpc-providers/README.md:
```diff
@@ -0,0 +1,19 @@
+## ProviderWithAgreement
+
+1. Works like `ProviderWithFallback, with an exception for two methods.
+2. `getBlockNumber` asks all providers for blockNumber and then picks `blockNumber.sort(ASC)[1]`
```
This text doesn't match with the implementation

koonopek replied:
Yeah, updated

hatskier commented on packages/rpc-providers/src/ProviderWithAgreement.ts:
```diff
@@ -0,0 +1,147 @@
+              }
+            }
+
+            results.set(currentResult, 1);
```
Wouldn't it overwrite the value to 1, even if before we set it to another value? I think it can cause problems with all expectedProvidersToAgree except for 2

koonopek replied:
Yes, I ate `else` statement very good catch :)

hatskier commented on packages/rpc-providers/src/ProviderWithFallback.ts:
```diff
@@ -0,0 +1,237 @@
+  }
+
+  private async doExecuteWithFallback(
+    retryNumber = 0,
```
I would call it alreadyRetriedCount as currently it's difficult to differentiate if from number of allowed retries

koonopek replied:
Definitely better name!

hatskier commented on packages/rpc-providers/src/ProviderWithFallback.ts:
```diff
@@ -0,0 +1,237 @@
+      throw error;
+    }
+
+    // we want to avoid situation were 2 calls fail in very short period
```
IMHO this comment is a bit misleading:
- I am not sure I understand this expression and not sure it's valid `this.providerIndex + 2 % this.providers.length`
- I would also shorten it to smth like this `// if another concurrent request already has changed provider we do nothing`

koonopek replied:
* I think it is valid I meant that we would jum by two providers ahead effectively instead of 1 (as it is happening now). But I have removed it because it make more trouble then help
* good point

koonopek commented on packages/rpc-providers/src/ProviderWithAgreement.ts:
```diff
@@ -0,0 +1,205 @@
+  blockNumbers: number[],
+  numberOfProviders: number
+): number => {
+  if (blockNumbers.length === 1) {
```
I tried but this just median i think it can't be simpler.
Remove first `if`

koonopek commented on packages/rpc-providers/src/ProviderWithAgreement.ts:
```diff
@@ -0,0 +1,205 @@
+      ...this.getProviderWithFallbackConfig(),
+    };
+    if (
+      this.agreementConfig.numberOfProvidersWhichHaveToAgree < 2 ||
```
done

koonopek commented on packages/rpc-providers/src/ProviderWithAgreement.ts:
```diff
@@ -0,0 +1,205 @@
+    return new Promise<string>((resolve, reject) => {
+      const electedBlockNumber = convertBlockTagToNumber(electedBlockTag);
+      const errors: Error[] = [];
+      const results = new Map<string, number>();
```
Why not it has `has` method which is cleaner then calling `obj['abc']` because what if `obj['abc']=0` it is false positive. I'm not sure if storing biggg strings in objects will work properly.
I didn't find any benchamrks in favour of Objects:
1. https://stackoverflow.com/questions/18541940/map-vs-object-in-javascript
2. https://www.zhenghao.io/posts/object-vs-map
3. https://bretcameron.medium.com/how-javascript-maps-can-make-your-code-faster-90f56bf61d9d (don't like the methodology here)

koonopek commented on packages/rpc-providers/src/ProviderWithAgreement.ts:
```diff
@@ -0,0 +1,205 @@
+            .getBlockNumber()
+            // ignore errors try again later
+            .catch();
+          if (stop) {
```
Good idea

koonopek commented on packages/rpc-providers/src/ProviderWithAgreement.ts:
```diff
@@ -0,0 +1,205 @@
+      const syncThenCall = async (providerIndex: number) => {
+        await sync(providerIndex);
+        await call(providerIndex);
+        handledResults += 1;
```
Meh, I don't like this syntax, but I don't mind :)

koonopek commented on packages/rpc-providers/src/ProviderWithAgreement.ts:
```diff
@@ -0,0 +1,205 @@
+    super(providers, config);
+    this.agreementConfig = {
+      ...defaultConfig,
+      ...this.getProviderWithFallbackConfig(),
```
done

koonopek commented on packages/rpc-providers/tsconfig.json:
```diff
@@ -0,0 +1,31 @@
+    "src/**/*.json",
+    "./artifacts/**/*.json"
+  ],
+  "files": [
```
Don't know some auto magic - fixed

koonopek commented on packages/rpc-providers/tsconfig.build.json:
```diff
@@ -0,0 +1,22 @@
+  "include": ["./src", "package.json", "./contracts", "src/**/*.json"],
+  "files": [
+    "./src/ProviderWithFallback.ts",
+    "./src/ProviderWithFallbackBase.ts"
```
Fixed

koonopek commented on packages/rpc-providers/README.md:
```diff
@@ -0,0 +1,24 @@
+
+1. If the provider fails on some operation new provider is elected.
+2. If all providers fail on the same operation error is thrown.
+3. When we switch to the next provider, all operations (excluding 1.) are executed by it till it fails.
```
Hmm good question

koonopek commented on packages/rpc-providers/src/common.ts:
```diff
@@ -0,0 +1,16 @@
+    new Promise(
+      (_r, reject) =>
+        (timer = setTimeout(
+          () => reject(`Timeout error ${timeoutMS} [MS]`),
```
oki doki

koonopek commented on packages/rpc-providers/src/ProviderWithFallback.ts:
```diff
@@ -0,0 +1,239 @@
+    ...args: any[]
+  ): Promise<any> {
+    try {
+      return await (this.currentProvider as any)[fnName](...[...args]);
```
It shallow copy args.
I think we could leave just `...args`

koonopek commented on packages/rpc-providers/src/ProviderWithFallback.ts:
```diff
@@ -0,0 +1,239 @@
+    const nextProviderIndex = (this.providerIndex + 1) % this.providers.length;
+    const nextProviderName = this.extractProviderName(nextProviderIndex);
+
+    logger.info(`Fallback in to next provider ${nextProviderName}.`);
```
Good idea

koonopek commented on packages/rpc-providers/test/ProviderWithFallback.test.ts:
```diff
@@ -0,0 +1,190 @@
+    expect(await fallbackProvider.getBlockNumber()).to.eq(10);
+  });
+
+  it("should call first provider if second fails, and the second provider is currently used", async () => {
```
Yep

koonopek commented on packages/rpc-providers/test/ProviderWithFallback.test.ts:
```diff
@@ -0,0 +1,190 @@
+      stubProvider,
+    ]);
+
+    spy.onFirstCall().rejects().onSecondCall().resolves(10);
```
Good idea, will do

koonopek commented on packages/rpc-providers/test/ProviderWithFallback.test.ts:
```diff
@@ -0,0 +1,190 @@
+  });
+});
+
+async function deployCounter() {
```
;)

hatskier commented on packages/rpc-providers/src/ProviderWithAgreement.ts:
```diff
@@ -0,0 +1,205 @@
+          blockPerProvider[providerIndex] = await this.providers[providerIndex]
+            .getBlockNumber()
+            // ignore errors try again later
+            .catch();
```
Hmm, I am still a bit afraid of never resolved or rejected promise if all rpcs just fail to return block number. They will never be synced but will always retry.

I would propose a small change, e.g. to stop trying to sync for a given rpc if one call has failed. Let's discuss it tomorrow
